### PR TITLE
Prune data

### PR DESCRIPTION
**Pull request by Siegrift**
Closes https://github.com/api3dao/signed-api/issues/35

## Rationale

I implemented it as a step during insert for performance and easy-ness to implement.  The pruning function takes a list of signed data for which it tries to remove no longer needed data. The pruning function is implemented as a separate function to reduce coupling between insertion and removal of the data.

### CODE REVIEW

File: packages/api/src/in-memory-cache.test.ts [hunk 1]
```diff
@@ -209,3 +209,37 @@ describe(putAll.name, () => {
     expect(cache[newDataBatch[1]!.airnode]![newDataBatch[1]!.templateId]).toEqual([newDataBatch[1]]);
   });
 });
+
+describe(prune.name, () => {
```
Comment (andreogle) [inline packages/api/src/in-memory-cache.test.ts:213]: I missed this before, pretty neat that you can do this 👍🏻

File: packages/api/src/in-memory-cache.ts [hunk 2]
```diff
@@ -54,3 +54,23 @@ export const put = async (signedData: SignedData) => {
 export const putAll = async (signedDataArray: SignedData[]) => {
   for (const signedData of signedDataArray) await put(signedData);
 };
+
+// The API is deliberately asynchronous to mimic a database call.
+//
+// Removes all signed data that is no longer needed to be kept in memory (because it is too old and there exist a newer
+// signed data for each endpoint). The function is intended to be called after each insertion of new signed data for
+// performance reasons, because it only looks to prune the data that for beacons that have been just inserted.
+export const prune = async (signedDataArray: SignedData[], maxIgnoreAfterTimestamp: number) => {
```
Comment (andreogle) [inline packages/api/src/in-memory-cache.ts:63]: Should the timestamp include the unit in the name? Or should it rather be a `Date` and the function handles converting to a Unix timestamp internally?

It's not obvious to me if "maxIgnoreAfterTimestamp" is in seconds or milliseconds and requires me to dig through several other functions to figure this out.

**Review by andreogle — APPROVED** [review]

File: packages/api/src/in-memory-cache.ts [hunk 3]
```diff
@@ -54,3 +54,23 @@ export const put = async (signedData: SignedData) => {
...
+  const beaconsToPrune = uniqBy(signedDataArray, 'beaconId');
+  const signedDataCache = getCache();
+
+  for (const beacon of beaconsToPrune) {
+    const { airnode, templateId } = beacon;
+    const allSignedDatas = signedDataCache[airnode]![templateId]!; // Assume the data is inserted the cache.
```
[the remark this answers is not in the document]
Comment (Siegrift) [reply packages/api/src/in-memory-cache.ts:69]: Yeah, but we always use `data` in singular and then there is the dillemma how to call a "group" of data. After a bit of internal pain, the "datas" was the best I could come up with. Calling it `allSignedDataArray` was second on my internal list afair. Do you like more?

File: packages/api/src/in-memory-cache.ts (hunk 2 above)
Comment (Siegrift) [reply packages/api/src/in-memory-cache.ts:63]: Timestamp should imo always refer to unix timestamp and thus be in seconds. We also use this implicitly for signed data currently (e.g. the input for the signed API expects the timestamp https://github.com/api3dao/signed-api/blob/044a9d71b1da335c368d89f6b0b1503fe0dad6b6/packages/api/src/schema.ts#L45 and assumes it's in seconds).

Also, my intention is for the cache to work with timestamps and not delays. That's why I made API handlers convert the required delay to timestamp before calling the cache functions.

File: packages/api/src/in-memory-cache.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreogle) [reply packages/api/src/in-memory-cache.ts:69]: I guess it's fine as is. Not sure how I feel about the alternative

[the remark this answers is not in the document]
Comment (bbenligiray) [reply packages/api/src/in-memory-cache.ts:69]: I also find `datas` useful (specifically `calldatas` in a multicall situation) but then https://github.com/api3dao/manager-multisig/pull/213#discussion_r1284344567

**Pull request merged**